### DISCUSSION TITLE
Can you create a schedule jobs solution?

### DISCUSSION THREAD

**Discussion by andrerocha** [jasontaylordev/CleanArchitecture#395] [Q&A]
I need create a schedule jobs. This is possible using the same structure?

Maybe use job worker or any other idea as quartz etc..

Anybody have any idea?

**Comment by neman** [comment]
Yes it is possible. I used background tasks, which executes commands from Application layer.

**Comment by litan1106** [comment]
I used Hangfire with Mediatr. it worked great.

**Comment by shobhitvaish** [comment]
> I used Hangfire with Mediatr. it worked great.

Mind sharing an example?

**Comment by MARKUSER** [comment]
1. I am looking for some help on the same. Using this project template i have started building the code. I have API project and at the same time i wanted to create .net core WorkerService (like windows application) to process CSV and EXCEL files. Any pointers to using this template with WorkerService will be of great help.
2. I would like to create some reusable logic between API and WorkerService so is it a good idea to create new libraries and call them from Application layer handlers ?

**Comment by MARKUSER** [comment]
> Yes it is possible. I used background tasks, which executes commands from Application layer.

@neman any help ?

**Comment by fakhrulhilal** [comment]
I never implement background job using clean architecture. But I ever implement background job using hangfire. Basically, we just register our action and let hangfire does the job. I'm thinking to create job executor abstraction in application layer, f.e 
```cs
public delegate Task JobAction();
public interface IJobAgent
{
    void Queue(JobAction action, JobConfigs configs);
}

public abstract record JobConfigs
{
    public static OneTimeJobConfigs OneTime => new();
    public static OneTimeJobConfigs Delayed(TimeSpan delay) => new(delay);
    public static RecurringJobConfigs Scheduled(string schedule) => new(schedule);
}
public record OneTimeJobConfigs(TimeSpan? Delay = null) : JobConfigs;
public record RecurringJobConfigs(string Schedule) : JobConfigs;

```
And then in infra layer, we can use hangfire like this:
```cs
public class HangfireJobAgent : IJobAgent
{
    public void Queue(JobAction action, JobConfigs configs)
    {
        switch (configs)
        {
            case RecurringJobConfigs recurringJob:
                RecurringJob.AddOrUpdate(() => action(), recurringJob.CronSchedule);
                break;
            case OneTimeJobConfigs { Delay: not null } oneTimeJob:
                BackgroundJob.Schedule(() => action(), oneTimeJob.Delay.Value);
                break;
            case OneTimeJobConfigs:
                BackgroundJob.Enqueue(() => action());
                break;
        }
    }
}

public static class DependencyInjection
{
    public static IServiceCollection AddAgentJob(this IServiceCollection services) => services
        // Add Hangfire services.
        .AddHangfire(configuration => configuration
            .SetDataCompatibilityLevel(CompatibilityLevel.Version_170)
            .UseSimpleAssemblyNameTypeSerializer()
            .UseRecommendedSerializerSettings()
            // choose your job storage options
            .UseSQLiteStorage("yourConnectionString"))

        // Add the processing server as IHostedService
        .AddHangfireServer();
}
```
After that, we can use in request handler as follows:
```cs
public record MyRequest : IRequest;
public class MyRequestHandler : IRequestHandler<MyRequest>
{
    private readonly IJobAgent _jobAgent;

    public MyRequestHandler(IJobAgent jobAgent)
    {
        _jobAgent = jobAgent;
    }

    public Task<Unit> Handle(MyRequest request, CancellationToken cancellationToken)
    {
        // register task next hour
        _jobAgent.Queue(() => Task.CompletedTask, JobConfigs.Delayed(TimeSpan.FromHours(1)));
        return Task.FromResult(Unit.Value);
    }
}
```

Note that, I never tried this, please test before using. This's just my idea.

**Comment by artworkad** [comment]
@fakhrulhilal thx for the example. Had to make following changes to make it work.

Method call expressions are not supported with Hangfire 

- https://github.com/HangfireIO/Hangfire/issues/30
- https://discuss.hangfire.io/t/generate-a-methodcallexpression-to-pass-to-hangfire-enqueue/1017

Minor changes to make it work:

```
using System.Linq.Expressions;

public interface IJobService
{
    void Queue<T>(Expression<Action<T>> action, JobConfigs configs);
}
```

```
public class HangfireJobService : IJobService
{
    public void Queue<T>(Expression<Action<T>> action, JobConfigs configs)
    {
        switch (configs)
        {
            case OneTimeJobConfigs:
                BackgroundJob.Enqueue(action);
                break;
            case RecurringJobConfigs recurringJob:
                RecurringJob.AddOrUpdate(action, recurringJob.Schedule);
                break;
        }
    }
}
```

Start a job from e.g. a query like so:

`_jobService.Queue<IReportService>(_reportService => _reportService .SomeMethod(request), JobConfigs.OneTime);`
